modules/ip/control/icmp.c

```diff
@@ -0,0 +1,146 @@
+// SPDX-License-Identifier: BSD-3-Clause
+// Copyright (c) 2024 Christophe Fontaine
+
+#include <gr_api.h>
+#include <gr_control.h>
+#include <gr_control_input.h>
+#include <gr_control_output.h>
+#include <gr_ip4.h>
+#include <gr_ip4_control.h>
+#include <gr_ip4_datapath.h>
+
+#include <rte_icmp.h>
+#include <rte_ip.h>
+
+#include <stdatomic.h>
+#include <time.h>
+
+#define ICMP_BUFFERED_REPLIES 32
+
+static struct rte_mbuf *replies[ICMP_BUFFERED_REPLIES];
+static uint16_t reply_index;
+
+// Global id which is used to differentiate between api clients
+static uint16_t icmp_ident;
+
+static struct rte_mbuf *ip4_icmp_get_reply(uint16_t id) {
+	struct rte_mbuf *null_mbuf = NULL;
+	struct rte_icmp_hdr *icmp;
+
+	for (int i = 0; i < ICMP_BUFFERED_REPLIES; i++) {
+		if (replies[i]) {
+			icmp = rte_pktmbuf_mtod(replies[i], struct rte_icmp_hdr *);
+			if (icmp->icmp_ident == htons(id))
+				return atomic_exchange(&replies[i], null_mbuf);
+		}
+	}
+	return NULL;
+}
+
+static void icmp_response(struct rte_mbuf *m) {
+	struct rte_mbuf *previous = atomic_exchange(&replies[reply_index++], m);
```

Comment:
Not sure if incrementing reply_index this way is safe.
Can't we use a little ring instead?

Reply:
will do.

Reply:
I used a preallocated array with stb_ds, hope this will do the job just as fine.